src/aktualizr_info/main.cc

```diff
@@ -77,219 +78,237 @@ int main(int argc, char **argv) {
+                                                                           &current_version, &pending_version);
+
+          if (!load_installed_version_res ||
+              (current_version >= installed_targets.size() && pending_version >= installed_targets.size())) {
+            std::cout << "   no details about installed nor pending images\n";
+          } else {
+            if (installed_targets.size() > current_version) {
+              std::cout << "   installed image hash: " << installed_targets[current_version].sha256Hash() << "\n";
+              std::cout << "   installed image filename: " << installed_targets[current_version].filename() << "\n";
+            }
+            if (installed_targets.size() > pending_version) {
+              std::cout << "   pending image hash: " << installed_targets[pending_version].sha256Hash() << "\n";
+              std::cout << "   pending image filename: " << installed_targets[pending_version].filename() << "\n";
+            }
           }
         }
       }
-    }
 
-    std::vector<MisconfiguredEcu> misconfigured_ecus;
-    storage->loadMisconfiguredEcus(&misconfigured_ecus);
-    if (misconfigured_ecus.size() != 0u) {
-      std::cout << "Removed or not registered ecus:" << std::endl;
-      std::vector<MisconfiguredEcu>::const_iterator it;
-      for (it = misconfigured_ecus.begin(); it != misconfigured_ecus.end(); ++it) {
-        std::cout << "   '" << it->serial << "' with hardware_id '" << it->hardware_id << "' "
-                  << (it->state == EcuState::kOld ? "has been removed from config" : "not registered yet") << std::endl;
+      std::vector<MisconfiguredEcu> misconfigured_ecus;
+      storage->loadMisconfiguredEcus(&misconfigured_ecus);
+      if (misconfigured_ecus.size() != 0u) {
+        std::cout << "Removed or not registered ecus:" << std::endl;
+        std::vector<MisconfiguredEcu>::const_iterator it;
+        for (it = misconfigured_ecus.begin(); it != misconfigured_ecus.end(); ++it) {
+          std::cout << "   '" << it->serial << "' with hardware_id '" << it->hardware_id << "' "
+                    << (it->state == EcuState::kOld ? "has been removed from config" : "not registered yet")
+                    << std::endl;
+        }
       }
-    }
 
-    auto registered = storage->loadEcuRegistered();
-    std::cout << "Provisioned on server: " << (registered ? "yes" : "no") << std::endl;
-    std::cout << "Fetched metadata: " << (has_metadata ? "yes" : "no") << std::endl;
+      auto registered = storage->loadEcuRegistered();
+      std::cout << "Provisioned on server: " << (registered ? "yes" : "no") << std::endl;
+      std::cout << "Fetched metadata: " << (has_metadata ? "yes" : "no") << std::endl;
+      provision_trigger = (provision_trigger && registered);
+      provision_trigger = (provision_trigger && has_metadata);
 
+      if (provision_trigger) {
+        break;
+      } else {
+        sleep(5);
```

Comment:
I think a much smaller sleep time would be better. 1 second is probably plenty. However, that will result in a lot of garbage printed to the screen, so we might want to any output until the loop finishes.

Does anyone else have opinions on what they'd like to see printed while waiting for provisioning to complete? Perhaps most ideal would be to only print the successful parts as they happen.

Reply:
I am just wondering if there is/are condition(s) under which this 'for' loop can last forever, e.g. device hasn't been registered at the server for some reason.

Reply:
I might have the same question as Mike said, under what condition(s) would the loop last forever? Now I just assume the conditions are :device ID and ECU serials could be loaded, device has been registered, metadata is available. I actually am not sure about it. Is provisioned means those four conditions are meet? @patrickvacek

Reply:
You can't register without device ID and ECU serials, and you can't fetch metadata without provisioning, so the only conditions that need to be checked are registration and metadata successfully fetched. It is true that this loop could last forever if something goes wrong. We've discussed a possible timeout feature, but for a first go of this feature, I don't think we need that yet.

Reply:
It probably shouldn't output anything until it's provisioned, so that the output can be parsed without errors / duplicates.

The body of the `while` loop looks quite big as well. Is it possible to have something like that?

```
bool provisioned = false;
while (wait_provisioning && !provisioned) {
  // some simple condition, no output
}

// regular aktualizr-info body
```

For the timeout parameter, as it's usually called as a subprocess by other tools (in python for example), we can use a timeout there for the moment.

Reply:
Thanks. Much clear now！

Reply:
> For the timeout parameter, as it's usually called as a subprocess by other tools (in python for example), we can use a timeout there for the moment.

So, how does the other tools（in python for example） outside  aktualizr-info main function body break this loop？

Reply:
For example, in Python:

https://docs.python.org/3/library/subprocess.html#subprocess.run

> The timeout argument is passed to Popen.communicate(). If the timeout expires, the child process will be killed and waited for. The TimeoutExpired exception will be re-raised after the child process has terminated.

Reply:
Ah, thanks!